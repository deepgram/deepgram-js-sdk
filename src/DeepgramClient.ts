import { AbstractClient } from "./packages/AbstractClient";
import { ListenClient } from "./packages/ListenClient";
import { ManageClient } from "./packages/ManageClient";
import { OnPremClient } from "./packages/OnPremClient";

/**
 * Deepgram Client.
 *
 * An isomorphic Javascript client for interacting with the Deepgram API.
 * @see https://developers.deepgram.com
 */
export default class DeepgramClient extends AbstractClient {
  get listen(): ListenClient {
    return new ListenClient(this.key, this.options);
  }

  get manage(): ManageClient {
    return new ManageClient(this.key, this.options);
  }

  get onprem(): OnPremClient {
    return new OnPremClient(this.key, this.options);
  }
}
