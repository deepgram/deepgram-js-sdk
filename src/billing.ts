import { BalanceList, Balance, RequestFunction } from "./types";

export class Billing {
  constructor(
    private _credentials: string,
    private _apiUrl: string,
    private _requireSSL: boolean,
    private _request: RequestFunction
  ) {}

  /**
   * Retrieves list of balance info of the specified project.
   * @param {string} projectId Unique identifier of the project
   * @param {string} endpoint Custom API endpoint
   *
   * @returns {Promise<BalanceList>}
   */
  async listBalances(
    projectId: string,
    endpoint = "v1/projects"
  ): Promise<BalanceList> {
    return this._request(
      "GET",
      this._credentials,
      this._apiUrl,
      this._requireSSL,
      `/${endpoint}/${projectId}/balances`
    );
  }

  /**
   * Retrieves balance info of a specified balance_id in the specified project.
   * @param {string} projectId Unique identifier of the project
   * @param {string} balanceId Unique identifier of the balance
   * @param {string} endpoint Custom API endpoint
   *
   * @returns {Promise<Balance>}
   */
  async getBalance(
    projectId: string,
    balanceId: string,
    endpoint = "v1/projects"
  ): Promise<Balance> {
    return this._request(
      "GET",
      this._credentials,
      this._apiUrl,
      this._requireSSL,
      `/${endpoint}/${projectId}/balances/${balanceId}`
    );
  }
}
