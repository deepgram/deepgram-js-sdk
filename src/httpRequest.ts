import { request, RequestOptions } from "https";
import { userAgent } from "./userAgent";

const _requestOptions = (
  api_key: string,
  apiUrl: string,
  path: string,
  method: string,
  payload?: string | Buffer,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  override_options?: any
): RequestOptions => {
  const options = {
    host: apiUrl,
    path,
    method,
    headers: {
      "User-Agent": userAgent(),
      "Content-Type": "application/json",
      Authorization: `token ${api_key}`,
      "Content-Length": payload ? Buffer.byteLength(payload) : undefined,
    },
  };
  if (payload === undefined) {
    delete options.headers["Content-Length"];
  }
  let headers = options.headers;
  if (override_options && override_options.headers) {
    headers = { ...headers, ...override_options.headers };
  }

  return { ...options, ...override_options, ...{ headers } };
};

export function _request<T>(
  method: string,
  api_key: string,
  apiUrl: string,
  path: string,
  payload?: string | Buffer,
  // eslint-disable-next-line @typescript-eslint/ban-types
  options?: Object
): Promise<T> {
  const requestOptions = _requestOptions(
    api_key,
    apiUrl,
    path,
    method,
    payload,
    options
  );
  return new Promise((resolve, reject) => {
    try {
      const httpRequest = request(requestOptions, (dgRes) => {
        let dgResContent = "";

        dgRes.on("data", (chunk) => {
          dgResContent += chunk;
        });

        dgRes.on("end", () => {
          let dgResponse;
          try {
            dgResponse = JSON.parse(dgResContent);
          } catch (err) {
            dgResponse = { error: dgResContent };
          }

          if (dgResponse.error) {
            reject(`DG: ${dgResContent}`);
          }
          resolve(dgResponse);
        });

        dgRes.on("error", (err) => {
          reject(`DG: ${err}`);
        });
      });

      httpRequest.on("error", (err) => {
        reject(`DG: ${err}`);
      });

      if (payload) {
        httpRequest.write(payload);
      }

      httpRequest.end();
    } catch (err) {
      reject(`DG: ${err}`);
    }
  });
}
