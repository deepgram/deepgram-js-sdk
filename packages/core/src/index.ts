import { DeepgramClientOptions, IKeyFactory } from "./lib/types/DeepgramClientOptions";
import { DeepgramVersionError } from "./lib/errors";
import DeepgramClient from "./DeepgramClient";

/**
 * This class is deprecated and should not be used. It throws a `DeepgramVersionError` when instantiated.
 *
 * @deprecated
 * @see https://dpgr.am/js-v3
 */
class Deepgram {
  constructor(protected apiKey: string, protected apiUrl?: string, protected requireSSL?: boolean) {
    throw new DeepgramVersionError();
  }
}

/**
 * Creates a new Deepgram client instance.
 *
 * @param {DeepgramClientArgs} args - Arguments to pass to the Deepgram client constructor.
 * @returns A new Deepgram client instance.
 */
function createClient(): DeepgramClient;
function createClient(key?: string | IKeyFactory): DeepgramClient;
function createClient(options?: DeepgramClientOptions): DeepgramClient;
function createClient(key?: string | IKeyFactory, options?: DeepgramClientOptions): DeepgramClient;
function createClient(
  keyOrOptions?: string | IKeyFactory | DeepgramClientOptions,
  options?: DeepgramClientOptions
): DeepgramClient {
  let resolvedOptions: DeepgramClientOptions = {};

  if (typeof keyOrOptions === "string" || typeof keyOrOptions === "function") {
    if (typeof options === "object") {
      resolvedOptions = options;
    }

    resolvedOptions.key = keyOrOptions;
  } else if (typeof keyOrOptions === "object") {
    resolvedOptions = keyOrOptions;
  }

  return new DeepgramClient(resolvedOptions);
}

export { createClient, DeepgramClient, Deepgram };

/**
 * Helpful exports.
 */
export * from "./packages";
export * from "./lib/types";
export * from "./lib/enums";
export * from "./lib/constants";
export * from "./lib/errors";
export * from "./lib/helpers";

/**
 * Captions. These will be tree-shaken if unused.
 *
 * @see https://github.com/deepgram/deepgram-node-captions
 *
 * import/export declarations don't do anything but set up an alias to the
 * exported variable, they do not count as a "use". Given their semantics,
 * they are tracked specially by any bundler and will not adversely affect
 * tree-shaking.
 */
export { webvtt, srt } from "@deepgram/captions";
