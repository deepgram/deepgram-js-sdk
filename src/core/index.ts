import { DeepgramVersionError } from "./lib/errors";
import DeepgramClient from "./DeepgramClient";

/**
 * This class is deprecated and should not be used. It throws a `DeepgramVersionError` when instantiated.
 *
 * @deprecated
 * @see https://dpgr.am/js-v3
 */
class Deepgram {
  constructor(protected apiKey: string, protected apiUrl?: string, protected requireSSL?: boolean) {
    throw new DeepgramVersionError();
  }
}

export { DeepgramClient, Deepgram };

/**
 * Core exports - all the low-level functionality
 */
export * from "./packages";
export * from "./lib/types";
export * from "./lib/enums";
export * from "./lib/constants";
export * from "./lib/errors";
export * from "./lib/helpers";
export * from "./lib/version";

/**
 * Captions. These will be tree-shaken if unused.
 *
 * @see https://github.com/deepgram/deepgram-node-captions
 */
export { webvtt, srt } from "@deepgram/captions";
