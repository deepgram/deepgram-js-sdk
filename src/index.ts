import DeepgramClient from "./DeepgramClient";
import { DeepgramVersionError } from "./lib/errors";
import type { DeepgramClientOptions } from "./lib/types";

/**
 * Major version fallback error
 */
class Deepgram {
  constructor(protected apiKey: string, protected apiUrl?: string, protected requireSSL?: boolean) {
    throw new DeepgramVersionError();
  }
}

/**
 * Creates a new Deepgram Client.
 */
const createClient = (apiKey: string, options: DeepgramClientOptions = {}): DeepgramClient => {
  return new DeepgramClient(apiKey, options);
};

export { createClient, DeepgramClient, Deepgram };

/**
 * Helpful exports.
 */
export * from "./packages";
export * from "./lib/types";
export * from "./lib/enums";
export * from "./lib/constants";
export * from "./lib/errors";
export * from "./lib/helpers";

/**
 * Captions. These will be tree-shaken if unused.
 *
 * @see https://github.com/deepgram/deepgram-node-captions
 *
 * import/export declarations don't do anything but set up an alias to the
 * exported variable, they do not count as a "use". Given their semantics,
 * they are tracked specially by any bundler and will not adversely affect
 * tree-shaking.
 */
export { webvtt, srt } from "@deepgram/captions";
