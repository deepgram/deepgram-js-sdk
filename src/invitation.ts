import { _request } from "./httpRequest";
import { Message, InvitationOptions, InvitationList } from "./types";

export class Invitation {
  constructor(private _credentials: string, private _apiUrl: string) {}

  private apiPath = "/v1/projects";

  /**
   * Lists all the current invites of a specified project.
   * @param projectId Unique identifier of the project
   */
  async list(projectId: string): Promise<InvitationList> {
    return _request<InvitationList>(
      "GET",
      this._credentials,
      this._apiUrl,
      `${this.apiPath}/${projectId}/invites`
    );
  }

  /**
   * Sends an invitation to join the specified project.
   * @param projectId Unique identifier of the project
   */
  async send(projectId: string, options: InvitationOptions): Promise<Message> {
    return _request<Message>(
      "POST",
      this._credentials,
      this._apiUrl,
      `${this.apiPath}/${projectId}/invites`,
      JSON.stringify({
        email: options.email,
        scope: options.scope,
      })
    );
  }

  /**
   * Removes the authenticated account from the specified project.
   * @param projectId Unique identifier of the project
   */
  async leave(projectId: string): Promise<Message> {
    return _request<Message>(
      "DELETE",
      this._credentials,
      this._apiUrl,
      `${this.apiPath}/${projectId}/leave`
    );
  }

  /**
   * Removes the specified email from the invitations on the specified project.
   * @param projectId Unique identifier of the project
   * @param email email address of the invitee
   * NOTE: This will return successful even if the email does not have an invite on the project.
   */
  async delete(projectId: string, email: string): Promise<Message> {
    return _request<Message>(
      "DELETE",
      this._credentials,
      this._apiUrl,
      `${this.apiPath}/${projectId}/invites/${email}`
    );
  }
}
