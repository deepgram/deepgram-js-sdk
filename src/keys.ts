import { _request } from "./httpRequest";
import { CreateKeyOptions, KeyResponse, Key } from "./types";

export class Keys {
  constructor(private _credentials: string, private _apiUrl: string) {}

  private apiPath = "/v1/projects";

  /**
   * Retrieves all keys associated with the provided projectId
   * @param projectId Unique identifier of the project containing API keys
   */
  async list(projectId: string): Promise<KeyResponse> {
    const response = await _request<KeyResponse>(
      "GET",
      this._credentials,
      this._apiUrl,
      `${this.apiPath}/${projectId}/keys`
    );

    const output = response.api_keys.map((apiKey) => {
      return {
        ...apiKey, 
        ...apiKey.api_key
      }
    });

    return { api_keys: output };
  }

  /**
   * Retrieves a specific key associated with the provided projectId
   * @param projectId Unique identifier of the project containing API keys
   * @param keyId Unique identifier for the key to retrieve
   */
  async get(projectId: string, keyId: string): Promise<Key> {
    return _request<Key>(
      "GET",
      this._credentials,
      this._apiUrl,
      `${this.apiPath}/${projectId}/keys/${keyId}`
    );
  }

  /**
   * Creates an API key with the provided scopes
   * @param projectId Unique identifier of the project to create an API key under
   * @param comment Comment to describe the key
   * @param scopes Permission scopes associated with the API key
   * @param options Optional options used when creating API keys
   */
  async create(
    projectId: string,
    comment: string,
    scopes: Array<string>,
    options?: CreateKeyOptions
  ): Promise<Key> {
    /** Throw an error if the user provided both expirationDate and timeToLive */
    if (
      options &&
      options.expirationDate !== undefined &&
      options.timeToLive !== undefined
    ) {
      throw new Error(
        "Please provide expirationDate or timeToLive or neither. Providing both is not allowed."
      );
    }

    return _request<Key>(
      "POST",
      this._credentials,
      this._apiUrl,
      `${this.apiPath}/${projectId}/keys`,
      JSON.stringify({
        comment,
        scopes,
        expiration_date:
          options && options.expirationDate
            ? options.expirationDate
            : undefined,
        time_to_live_in_seconds:
          options && options.timeToLive ? options.timeToLive : undefined,
      })
    );
  }

  /**
   * Deletes an API key
   * @param projectId Unique identifier of the project to create an API key under
   * @param keyId Unique identifier for the key to delete
   */
  async delete(projectId: string, keyId: string): Promise<void> {
    return _request<void>(
      "DELETE",
      this._credentials,
      this._apiUrl,
      `${this.apiPath}/${projectId}/keys/${keyId}`
    );
  }
}
