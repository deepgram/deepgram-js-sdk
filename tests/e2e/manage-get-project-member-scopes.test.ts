import { createClient } from "../../src/index";
import { structureOnlySerializer, setupApiMocks, cleanupApiMocks } from "../__utils__";
import { testProjectIds, testMemberIds } from "../__fixtures__/manage";
import type { DeepgramResponse, GetProjectMemberScopesResponse } from "../../src/core/lib/types";

describe("manage getProjectMemberScopes E2E", () => {
  let deepgram: ReturnType<typeof createClient>;

  beforeAll(() => {
    // Set up API mocks (only active when not updating snapshots)
    setupApiMocks();

    // Use mock API key when mocking, real one when updating snapshots
    const apiKey = process.env.DEEPGRAM_API_KEY || "mock-api-key";
    deepgram = createClient(apiKey);

    // Add our custom serializer
    expect.addSnapshotSerializer(structureOnlySerializer);
  });

  afterAll(() => {
    // Clean up mocks
    cleanupApiMocks();
  });

  it("should retrieve member scopes for a specific project member", async () => {
    const response: DeepgramResponse<GetProjectMemberScopesResponse> =
      await deepgram.manage.getProjectMemberScopes(testProjectIds.primary, testMemberIds.member1);

    expect(response.error).toBeNull();
    expect(response.result).toBeDefined();

    if (response.result) {
      expect(response.result).toHaveProperty("scopes");
      expect(Array.isArray(response.result.scopes)).toBe(true);

      // Verify scope values if any scopes exist
      if (response.result.scopes.length > 0) {
        response.result.scopes.forEach((scope) => {
          expect(typeof scope).toBe("string");
        });
      }
    }

    expect(response).toMatchSnapshot("manage-getProjectMemberScopes-response-structure");
  });

  it("should handle custom endpoint for getProjectMemberScopes", async () => {
    const customEndpoint = `:version/projects/${testProjectIds.primary}/members/${testMemberIds.member1}/scopes`;
    const response: DeepgramResponse<GetProjectMemberScopesResponse> =
      await deepgram.manage.getProjectMemberScopes(
        testProjectIds.primary,
        testMemberIds.member1,
        customEndpoint
      );

    expect(response.error).toBeNull();
    expect(response.result).toBeDefined();

    if (response.result) {
      expect(response.result.scopes).toBeDefined();
      expect(Array.isArray(response.result.scopes)).toBe(true);
    }
  }, 30000); // 30 second timeout for API call
});
