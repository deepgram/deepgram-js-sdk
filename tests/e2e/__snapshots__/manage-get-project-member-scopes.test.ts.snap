// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`manage getProjectMemberScopes E2E should retrieve member scopes for a specific project member: manage-getProjectMemberScopes-response-structure 1`] = `
{
  "error": null,
  "result": {
    "scopes": [
      "<string>"
    ]
  }
}
`;
